File: app/src/main/java/org/tasks/caldav/BaseCaldavAccountSettingsActivity.java
package org.tasks.caldav;

import static com.todoroo.astrid.data.Task.NO_ID;
import static org.tasks.Strings.isNullOrEmpty;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;
import android.text.util.Linkify;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.appcompat.widget.Toolbar;
import at.bitfire.dav4jvm.exception.HttpException;
import butterknife.ButterKnife;
import butterknife.OnFocusChange;
import butterknife.OnTextChanged;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.todoroo.astrid.service.TaskDeleter;
import java.net.ConnectException;
import java.net.IDN;
import java.net.URI;
import java.net.URISyntaxException;
import javax.inject.Inject;
import org.tasks.R;
import org.tasks.billing.Inventory;
import org.tasks.billing.PurchaseActivity;
import org.tasks.data.CaldavAccount;
import org.tasks.data.CaldavDao;
import org.tasks.databinding.ActivityCaldavAccountSettingsBinding;
import org.tasks.dialogs.DialogBuilder;
import org.tasks.injection.ThemedInjectingAppCompatActivity;
import org.tasks.security.KeyStoreEncryption;
import org.tasks.ui.DisplayableException;
import timber.log.Timber;

public abstract class BaseCaldavAccountSettingsActivity extends ThemedInjectingAppCompatActivity
    implements Toolbar.OnMenuItemClickListener {

  public static final String EXTRA_CALDAV_DATA = "caldavData"; // $NON-NLS-1$
  protected static final String PASSWORD_MASK = "\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022";
  @Inject protected CaldavDao caldavDao;
  @Inject protected KeyStoreEncryption encryption;
  @Inject DialogBuilder dialogBuilder;
  @Inject TaskDeleter taskDeleter;
  @Inject Inventory inventory;

  protected CaldavAccount caldavAccount;

  protected ActivityCaldavAccountSettingsBinding binding;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    binding = ActivityCaldavAccountSettingsBinding.inflate(getLayoutInflater());
    setContentView(binding.getRoot());
    ButterKnife.bind(this);

    caldavAccount =
        savedInstanceState == null
            ? getIntent().getParcelableExtra(EXTRA_CALDAV_DATA)
            : savedInstanceState.getParcelable(EXTRA_CALDAV_DATA);

    if (caldavAccount == null || caldavAccount.getId() == NO_ID) {
      binding.nameLayout.setVisibility(View.GONE);
      binding.description.setVisibility(View.VISIBLE);
      binding.description.setText(getDescription());
      Linkify.addLinks(binding.description, Linkify.WEB_URLS);
    } else {
      binding.nameLayout.setVisibility(View.VISIBLE);
      binding.description.setVisibility(View.GONE);
    }

    if (savedInstanceState == null) {
      if (caldavAccount != null) {
        binding.name.setText(caldavAccount.getName());
        binding.url.setText(caldavAccount.getUrl());
        binding.user.setText(caldavAccount.getUsername());
        if (!isNullOrEmpty(caldavAccount.getPassword())) {
          binding.password.setText(PASSWORD_MASK);
        }
        binding.repeat.setChecked(caldavAccount.isSuppressRepeatingTasks());
      }
    }

    Toolbar toolbar = binding.toolbar.toolbar;

    toolbar.setTitle(
        caldavAccount == null ? getString(R.string.add_account) : caldavAccount.getName());
    toolbar.setNavigationIcon(getDrawable(R.drawable.ic_outline_save_24px));
    toolbar.setNavigationOnClickListener(v -> save());
    toolbar.inflateMenu(R.menu.menu_caldav_account_settings);
    toolbar.setOnMenuItemClickListener(this);
    toolbar.showOverflowMenu();
    themeColor.apply(toolbar);

    if (caldavAccount == null) {
      toolbar.getMenu().findItem(R.id.remove).setVisible(false);
      binding.name.requestFocus();
      InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
      imm.showSoftInput(binding.name, InputMethodManager.SHOW_IMPLICIT);
    }

    if (!inventory.hasPro()) {
      newSnackbar(getString(R.string.this_feature_requires_a_subscription))
          .setDuration(BaseTransientBottomBar.LENGTH_INDEFINITE)
          .setAction(
              R.string.button_subscribe,
              v -> startActivity(new Intent(this, PurchaseActivity.class)))
          .show();
    }
  }

  protected abstract @StringRes int getDescription();

  @Override
  protected void onSaveInstanceState(@NonNull Bundle outState) {
    super.onSaveInstanceState(outState);

    outState.putParcelable(EXTRA_CALDAV_DATA, caldavAccount);
  }

  private void showProgressIndicator() {
    binding.progressBar.progressBar.setVisibility(View.VISIBLE);
  }

  protected void hideProgressIndicator() {
    binding.progressBar.progressBar.setVisibility(View.GONE);
  }

  private boolean requestInProgress() {
    return binding.progressBar.progressBar.getVisibility() == View.VISIBLE;
  }

  @OnTextChanged(R.id.name)
  void onNameChanged() {
    binding.nameLayout.setError(null);
  }

  @OnTextChanged(R.id.url)
  void onUrlChanged() {
    binding.urlLayout.setError(null);
  }

  @OnTextChanged(R.id.user)
  void onUserChanged() {
    binding.userLayout.setError(null);
  }

  @OnTextChanged(R.id.password)
  void onPasswordChanged() {
    binding.passwordLayout.setError(null);
  }

  @OnFocusChange(R.id.password)
  void onPasswordFocused(boolean hasFocus) {
    if (hasFocus) {
      if (PASSWORD_MASK.equals(binding.password.getText().toString())) {
        binding.password.setText("");
      }
    } else {
      if (TextUtils.isEmpty(binding.password.getText()) && caldavAccount != null) {
        binding.password.setText(PASSWORD_MASK);
      }
    }
  }

  protected String getNewName() {
    String name = binding.name.getText().toString().trim();
    return isNullOrEmpty(name) ? getNewUsername() : name;
  }

  protected String getNewURL() {
    return binding.url.getText().toString().trim();
  }

  protected String getNewUsername() {
    return binding.user.getText().toString().trim();
  }

  boolean passwordChanged() {
    return caldavAccount == null || !PASSWORD_MASK.equals(binding.password.getText().toString().trim());
  }

  protected abstract String getNewPassword();

  private void save() {
    if (requestInProgress()) {
      return;
    }

    String username = getNewUsername();
    String url = getNewURL();
    String password = getNewPassword();

    boolean failed = false;

    if (isNullOrEmpty(url)) {
      binding.urlLayout.setError(getString(R.string.url_required));
      failed = true;
    } else {
      Uri baseURL = Uri.parse(url);
      String scheme = baseURL.getScheme();
      if ("https".equalsIgnoreCase(scheme) || "http".equalsIgnoreCase(scheme)) {
        String host = baseURL.getHost();
        if (isNullOrEmpty(host)) {
          binding.urlLayout.setError(getString(R.string.url_host_name_required));
          failed = true;
        } else {
          try {
            host = IDN.toASCII(host);
          } catch (Exception e) {
            Timber.e(e);
          }
          String path = baseURL.getEncodedPath();
          int port = baseURL.getPort();
          try {
            new URI(scheme, null, host, port, path, null, null);
          } catch (URISyntaxException e) {
            binding.urlLayout.setError(e.getLocalizedMessage());
            failed = true;
          }
        }
      } else {
        binding.urlLayout.setError(getString(R.string.url_invalid_scheme));
        failed = true;
      }
    }

    if (isNullOrEmpty(username)) {
      binding.userLayout.setError(getString(R.string.username_required));
      failed = true;
    }

    if (isNullOrEmpty(password)) {
      binding.passwordLayout.setError(getString(R.string.password_required));
      failed = true;
    }

    if (failed) {
      return;
    }

    if (caldavAccount == null) {
      showProgressIndicator();
      addAccount(url, username, password);
    } else if (needsValidation()) {
      showProgressIndicator();
      updateAccount(url, username, password);
    } else if (hasChanges()) {
      updateAccount();
    } else {
      finish();
    }
  }

  protected abstract void addAccount(String url, String username, String password);

  protected abstract void updateAccount(String url, String username, String password);

  protected abstract void updateAccount();

  protected abstract String getHelpUrl();

  protected void requestFailed(Throwable t) {
    hideProgressIndicator();

    if (t instanceof HttpException) {
      if (((HttpException) t).getCode() == 401) {
        showSnackbar(R.string.invalid_username_or_password);
      } else {
        showSnackbar(t.getMessage());
      }
    } else if (t instanceof DisplayableException) {
      showSnackbar(((DisplayableException) t).getResId());
    } else if (t instanceof ConnectException) {
      showSnackbar(R.string.network_error);
    } else {
      Timber.e(t);
      showSnackbar(R.string.error_adding_account, t.getMessage());
    }
  }

  private void showSnackbar(int resId, Object... formatArgs) {
    showSnackbar(getString(resId, formatArgs));
  }

  private void showSnackbar(String message) {
    newSnackbar(message).show();
  }

  private Snackbar newSnackbar(String message) {
    Snackbar snackbar =
        Snackbar.make(binding.rootLayout, message, 8000)
            .setTextColor(getColor(R.color.snackbar_text_color))
            .setActionTextColor(getColor(R.color.snackbar_action_color));
    snackbar
        .getView()
        .setBackgroundColor(getColor(R.color.snackbar_background));
    return snackbar;
  }

  private boolean hasChanges() {
    if (caldavAccount == null) {
      return !isNullOrEmpty(binding.name.getText().toString().trim())
          || !isNullOrEmpty(getNewPassword())
          || !isNullOrEmpty(binding.url.getText().toString().trim())
          || !isNullOrEmpty(getNewUsername())
          || binding.repeat.isChecked();
    }
    return needsValidation()
        || !getNewName().equals(caldavAccount.getName())
        || binding.repeat.isChecked() != caldavAccount.isSuppressRepeatingTasks();
  }

  protected boolean needsValidation() {
    return !getNewURL().equals(caldavAccount.getUrl())
        || !getNewUsername().equals(caldavAccount.getUsername())
        || passwordChanged();
  }

  @Override
  public void finish() {
    if (!requestInProgress()) {
      InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
      imm.hideSoftInputFromWindow(binding.name.getWindowToken(), 0);
      super.finish();
    }
  }

  @Override
  public void onBackPressed() {
    discard();
  }

  private void removeAccountPrompt() {
    if (requestInProgress()) {
      return;
    }

    dialogBuilder
        .newDialog()
        .setMessage(R.string.logout_warning, caldavAccount.getName())
        .setPositiveButton(R.string.remove, (dialog, which) -> removeAccount())
        .setNegativeButton(android.R.string.cancel, null)
        .show();
  }

  protected void removeAccount() {
    taskDeleter.delete(caldavAccount);
    setResult(RESULT_OK);
    finish();
  }

  private void discard() {
    if (requestInProgress()) {
      return;
    }

    if (!hasChanges()) {
      finish();
    } else {
      dialogBuilder
          .newDialog(R.string.discard_changes)
          .setPositiveButton(R.string.discard, (dialog, which) -> finish())
          .setNegativeButton(android.R.string.cancel, null)
          .show();
    }
  }

  @Override
  public boolean onMenuItemClick(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.menu_help:
        startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(getHelpUrl())));
        break;
      case R.id.remove:
        removeAccountPrompt();
        break;
    }
    return onOptionsItemSelected(item);
  }
}

File: app/src/main/java/org/tasks/caldav/BaseCaldavAccountSettingsActivity.kt
package org.tasks.caldav

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import androidx.annotation.StringRes
import androidx.appcompat.content.res.AppCompatResources
import androidx.appcompat.widget.Toolbar
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.core.content.ContextCompat
import androidx.core.widget.addTextChangedListener
import androidx.lifecycle.lifecycleScope
import at.bitfire.dav4jvm.exception.HttpException
import com.franmontiel.persistentcookiejar.persistence.CookiePersistor
import com.google.android.material.composethemeadapter.MdcTheme
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import com.todoroo.astrid.data.Task
import com.todoroo.astrid.service.TaskDeleter
import kotlinx.coroutines.launch
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import org.tasks.R
import org.tasks.Strings.isNullOrEmpty
import org.tasks.analytics.Firebase
import org.tasks.billing.Inventory
import org.tasks.billing.PurchaseActivity
import org.tasks.compose.ServerSelector
import org.tasks.data.CaldavAccount
import org.tasks.data.CaldavAccount.Companion.SERVER_UNKNOWN
import org.tasks.data.CaldavDao
import org.tasks.databinding.ActivityCaldavAccountSettingsBinding
import org.tasks.dialogs.DialogBuilder
import org.tasks.dialogs.Linkify
import org.tasks.extensions.Context.cookiePersistor
import org.tasks.extensions.Context.hideKeyboard
import org.tasks.extensions.Context.openUri
import org.tasks.extensions.addBackPressedCallback
import org.tasks.injection.ThemedInjectingAppCompatActivity
import org.tasks.security.KeyStoreEncryption
import org.tasks.ui.DisplayableException
import timber.log.Timber
import java.net.ConnectException
import java.net.IDN
import java.net.URI
import java.net.URISyntaxException
import javax.inject.Inject

abstract class BaseCaldavAccountSettingsActivity : ThemedInjectingAppCompatActivity(), Toolbar.OnMenuItemClickListener {
    @Inject lateinit var caldavDao: CaldavDao
    @Inject lateinit var encryption: KeyStoreEncryption
    @Inject lateinit var dialogBuilder: DialogBuilder
    @Inject lateinit var taskDeleter: TaskDeleter
    @Inject lateinit var inventory: Inventory
    @Inject lateinit var firebase: Firebase

    protected var caldavAccount: CaldavAccount? = null
    protected lateinit var binding: ActivityCaldavAccountSettingsBinding
    protected lateinit var serverType: MutableState<Int>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCaldavAccountSettingsBinding.inflate(layoutInflater)
        setContentView(binding.root)
        caldavAccount = if (savedInstanceState == null) intent.getParcelableExtra(EXTRA_CALDAV_DATA) else savedInstanceState.getParcelable(EXTRA_CALDAV_DATA)
        serverType = mutableStateOf(
            savedInstanceState?.getInt(EXTRA_SERVER_TYPE, SERVER_UNKNOWN)
                ?: caldavAccount?.serverType
                ?: SERVER_UNKNOWN
        )
        if (caldavAccount == null || caldavAccount!!.id == Task.NO_ID) {
            binding.nameLayout.visibility = View.GONE
            binding.description.visibility = View.VISIBLE
            binding.description.setText(description)
            Linkify.safeLinkify(binding.description, android.text.util.Linkify.WEB_URLS)
            serverType.value = SERVER_UNKNOWN
        } else {
            binding.nameLayout.visibility = View.VISIBLE
            binding.description.visibility = View.GONE
            caldavAccount?.error?.takeIf { it.isNotBlank() }?.let {
                binding.description.visibility = View.VISIBLE
                binding.description.setTextColor(ContextCompat.getColor(this, R.color.overdue))
                binding.description.text = getString(R.string.error_adding_account, it)
            }
        }
        if (savedInstanceState == null) {
            caldavAccount?.let {
                binding.name.setText(it.name)
                binding.url.setText(it.url)
                binding.user.setText(it.username)
                if (!isNullOrEmpty(it.password)) {
                    binding.password.setText(PASSWORD_MASK)
                }
                serverType.value = it.serverType
            }
        }
        val toolbar = binding.toolbar.toolbar
        toolbar.title = if (caldavAccount == null) getString(R.string.add_account) else caldavAccount!!.name
        toolbar.navigationIcon = AppCompatResources.getDrawable(this, R.drawable.ic_outline_save_24px)
        toolbar.setNavigationOnClickListener { save() }
        toolbar.inflateMenu(menuRes)
        toolbar.setOnMenuItemClickListener(this)
        toolbar.showOverflowMenu()
        if (caldavAccount == null) {
            toolbar.menu.findItem(R.id.remove).isVisible = false
            binding.name.requestFocus()
            val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.showSoftInput(binding.name, InputMethodManager.SHOW_IMPLICIT)
        }
        if (!inventory.hasPro) {
            newSnackbar(getString(R.string.this_feature_requires_a_subscription))
                    .setDuration(BaseTransientBottomBar.LENGTH_INDEFINITE)
                    .setAction(R.string.button_subscribe) {
                        startActivity(Intent(this, PurchaseActivity::class.java))
                    }
                    .show()
        }
        binding.name.addTextChangedListener(
            onTextChanged = { _, _, _, _ -> binding.nameLayout.error = null }
        )
        binding.url.addTextChangedListener(
            onTextChanged = { _, _, _, _ -> binding.urlLayout.error = null }
        )
        binding.user.addTextChangedListener(
            onTextChanged = { _, _, _, _ -> binding.userLayout.error = null }
        )
        binding.password.addTextChangedListener(
            onTextChanged = { _, _, _, _ -> binding.passwordLayout.error = null }
        )
        binding.password.setOnFocusChangeListener { _, hasFocus -> onPasswordFocused(hasFocus) }
        binding.serverSelector.setContent {
            MdcTheme {
                var selected by rememberSaveable { serverType }
                ServerSelector(selected) {
                    serverType.value = it
                    selected = it
                }
            }
        }

        addBackPressedCallback {
            discard()
        }
    }

    @get:StringRes
    protected open val description = 0

    protected open val menuRes = R.menu.menu_caldav_account_settings

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(EXTRA_CALDAV_DATA, caldavAccount)
        outState.putInt(EXTRA_SERVER_TYPE, serverType.value)
    }

    private fun showProgressIndicator() {
        binding.progressBar.progressBar.visibility = View.VISIBLE
    }

    protected fun hideProgressIndicator() {
        binding.progressBar.progressBar.visibility = View.GONE
    }

    private fun requestInProgress(): Boolean {
        return binding.progressBar.progressBar.visibility == View.VISIBLE
    }

    private fun onPasswordFocused(hasFocus: Boolean) {
        if (hasFocus) {
            if (PASSWORD_MASK == binding.password.text.toString()) {
                binding.password.setText("")
            }
        } else if (TextUtils.isEmpty(binding.password.text) && caldavAccount != null) {
            binding.password.setText(PASSWORD_MASK)
        }
    }

    protected val newName: String
        get() {
            val name = binding.name.text.toString().trim { it <= ' ' }
            return if (isNullOrEmpty(name)) newUsername else name
        }

    protected open val newURL: String
        get() = binding.url.text.toString().trim { it <= ' ' }

    protected val newUsername: String
        get() = binding.user.text.toString().trim { it <= ' ' }

    fun passwordChanged(): Boolean {
        return caldavAccount == null || PASSWORD_MASK != binding.password.text.toString().trim { it <= ' ' }
    }

    protected abstract val newPassword: String?

    protected open fun save() = lifecycleScope.launch {
        if (requestInProgress()) {
            return@launch
        }
        val username = newUsername
        val url = newURL
        val password = newPassword
        var failed = false
        if (newName.isBlank()) {
            binding.nameLayout.error = getString(R.string.name_cannot_be_empty)
            failed = true
        }
        if (isNullOrEmpty(url)) {
            binding.urlLayout.error = getString(R.string.url_required)
            failed = true
        } else {
            val baseURL = Uri.parse(url)
            val scheme = baseURL.scheme
            if ("https".equals(scheme, ignoreCase = true) || "http".equals(scheme, ignoreCase = true)) {
                var host = baseURL.host
                if (isNullOrEmpty(host)) {
                    binding.urlLayout.error = getString(R.string.url_host_name_required)
                    failed = true
                } else {
                    try {
                        host = IDN.toASCII(host)
                    } catch (e: Exception) {
                        Timber.e(e)
                    }
                    val path = baseURL.encodedPath
                    val port = baseURL.port
                    try {
                        URI(scheme, null, host, port, path, null, null)
                    } catch (e: URISyntaxException) {
                        binding.urlLayout.error = e.localizedMessage
                        failed = true
                    }
                }
            } else {
                binding.urlLayout.error = getString(R.string.url_invalid_scheme)
                failed = true
            }
        }
        if (isNullOrEmpty(username)) {
            binding.userLayout.error = getString(R.string.username_required)
            failed = true
        }
        if (isNullOrEmpty(password)) {
            binding.passwordLayout.error = getString(R.string.password_required)
            failed = true
        }
        when {
            failed -> return@launch
            caldavAccount == null -> {
                showProgressIndicator()
                addAccount(url, username, password!!)
            }
            needsValidation() -> {
                showProgressIndicator()
                updateAccount(url, username, password!!)
            }
            hasChanges() -> {
                updateAccount()
            }
            else -> {
                finish()
            }
        }
    }

    protected abstract suspend fun addAccount(url: String, username: String, password: String)
    protected abstract suspend fun updateAccount(url: String, username: String, password: String)
    protected abstract suspend fun updateAccount()
    protected abstract val helpUrl: Int

    protected fun requestFailed(t: Throwable) {
        hideProgressIndicator()
        when (t) {
            is HttpException ->
                if (t.code == 401)
                    showSnackbar(R.string.invalid_username_or_password)
                else
                    showSnackbar(t.message)
            is DisplayableException -> showSnackbar(t.resId)
            is ConnectException -> showSnackbar(R.string.network_error)
            else -> {
                Timber.e(t)
                showSnackbar(R.string.error_adding_account, t.message!!)
            }
        }
    }

    private fun showSnackbar(resId: Int, vararg formatArgs: Any) {
        showSnackbar(getString(resId, *formatArgs))
    }

    private fun showSnackbar(message: String?) {
        newSnackbar(message).show()
    }

    private fun newSnackbar(message: String?): Snackbar {
        val snackbar = Snackbar.make(binding.rootLayout, message!!, 8000)
                .setTextColor(getColor(R.color.snackbar_text_color))
                .setActionTextColor(getColor(R.color.snackbar_action_color))
        snackbar
                .view
                .setBackgroundColor(getColor(R.color.snackbar_background))
        return snackbar
    }

    protected open fun hasChanges(): Boolean {
        return if (caldavAccount == null) {
            (!isNullOrEmpty(binding.name.text.toString().trim { it <= ' ' })
                    || !isNullOrEmpty(newPassword)
                    || !isNullOrEmpty(binding.url.text.toString().trim { it <= ' ' })
                    || !isNullOrEmpty(newUsername)
                    || serverType.value != SERVER_UNKNOWN
            )
        } else needsValidation() ||
                newName != caldavAccount!!.name ||
                serverType.value != caldavAccount!!.serverType
    }

    protected open fun needsValidation(): Boolean =
            newURL != caldavAccount!!.url
                    || newUsername != caldavAccount!!.username
                    || passwordChanged()

    override fun finish() {
        if (!requestInProgress()) {
            hideKeyboard(binding.name)
            super.finish()
        }
    }

    private fun removeAccountPrompt() {
        if (requestInProgress()) {
            return
        }
        dialogBuilder
                .newDialog()
                .setMessage(R.string.logout_warning)
                .setPositiveButton(R.string.remove) { _, _ -> lifecycleScope.launch { removeAccount() } }
                .setNegativeButton(R.string.cancel, null)
                .show()
    }

    protected open suspend fun removeAccount() {
        cookiePersistor(caldavAccount?.username).clearSession(caldavAccount?.url)
        taskDeleter.delete(caldavAccount!!)
        setResult(Activity.RESULT_OK)
        finish()
    }

    private fun discard() {
        if (requestInProgress()) {
            return
        }
        if (hasChanges()) {
            dialogBuilder
                    .newDialog(R.string.discard_changes)
                    .setPositiveButton(R.string.discard) { _, _ -> finish() }
                    .setNegativeButton(R.string.cancel, null)
                    .show()
        } else {
            finish()
        }
    }

    override fun onMenuItemClick(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_help -> openUri(helpUrl)
            R.id.remove -> removeAccountPrompt()
        }
        return onOptionsItemSelected(item)
    }

    companion object {
        const val EXTRA_CALDAV_DATA = "caldavData" // $NON-NLS-1$
        const val EXTRA_SERVER_TYPE = "serverType"
        const val PASSWORD_MASK = "\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022"

        fun CookiePersistor.clearSession(url: String?) {
            val httpUrl = url?.toHttpUrlOrNull() ?: return
            removeAll(loadAll().filter { it.matches(httpUrl) })
        }
    }
}